--- a/setup.py
+++ b/setup.py
@@ -503,9 +503,9 @@
             ] + glob.glob("apex/contrib/csrc/group_norm/*.cu"),
             include_dirs=[os.path.join(this_dir, 'csrc')],
             extra_compile_args={
-                "cxx": ["-O3", "-std=c++17"],
-                "nvcc": [
-                    "-O3", "-std=c++17", "--use_fast_math", "--ftz=false",
+                "cxx": ["-O3", "-std=c++20"],
+                "nvcc": [
+                    "-O3", "-std=c++20", "--use_fast_math", "--ftz=false",
                 ],
             },
         )
