--- a/setup.py
+++ b/setup.py
@@ -186,6 +186,9 @@
         if bare_metal_version >= Version("12.8") and "100" in cuda_archs():
             cc_flag.append("-gencode")
             cc_flag.append("arch=compute_100,code=sm_100")
+        if bare_metal_version >= Version("13.4") and "107" in cuda_archs():
+            cc_flag.append("-gencode")
+            cc_flag.append("arch=compute_107,code=sm_107")
         if bare_metal_version >= Version("12.8") and "120" in cuda_archs():
             cc_flag.append("-gencode")
             cc_flag.append("arch=compute_120,code=sm_120")
@@ -199,6 +202,6 @@
     nvcc_flags = [
     "-O3",
-    "-std=c++17",
+    "-std=c++20",
     "-U__CUDA_NO_HALF_OPERATORS__",
     "-U__CUDA_NO_HALF_CONVERSIONS__",
     "-U__CUDA_NO_HALF2_OPERATORS__",
@@ -220,4 +223,4 @@
-    compiler_c17_flag=["-O3", "-std=c++17"]
+    compiler_c17_flag=["-O3", "-std=c++20"]
     # Add Windows-specific flags
     if sys.platform == "win32" and os.getenv('DISTUTILS_USE_SDK') == '1':
         nvcc_flags.extend(["-Xcompiler", "/Zc:__cplusplus"])
