diff --git a/python/sglang/srt/layers/attention/tbo_backend.py b/python/sglang/srt/layers/attention/tbo_backend.py
index 494d82d80..2ae120686 100644
--- a/python/sglang/srt/layers/attention/tbo_backend.py
+++ b/python/sglang/srt/layers/attention/tbo_backend.py
@@ -179,6 +179,9 @@ class TboAttnBackend(AttentionBackend):
             assert ans == child.get_cuda_graph_seq_len_fill_value()
         return ans
 
+    def forward(self, *args, **kwargs):
+        return self.primary.forward(*args, **kwargs)
+
     def forward_extend(self, *args, **kwargs):
         return self.primary.forward_extend(*args, **kwargs)
 
