--- megatron
+++ hook_rollout
@@ -17,0 +18,2 @@
+- --rollout-function-path
+- tests.fast.argument_snapshot.argparser.scenarios._Hook
@@ -1020 +1022 @@
-    rollout_function_path: null
+    rollout_function_path: tests.fast.argument_snapshot.argparser.scenarios._Hook
@@ -1675,0 +1678 @@
+    snapshot_hook: 23
@@ -26809,0 +26813,8 @@
+      - default: 23
+        dest: snapshot_hook
+        group: options
+        help: null
+        option_strings:
+        - --snapshot-hook
+        type:
+          $callable: builtins.int
